accepted/00065-salt-minions-update-cycle.md

```diff
@@ -0,0 +1,127 @@
+
+Consequently, this mechanism will allow to keep always aligned Salt Minions across the entire infrastructure consistent and identical.
+
+## Fail-proof Updates
+
+Having running Salt Minion inside a virtual environment, allows to deliver fail-proof updates. The basic mechanism is to clone Python Virtual Environment from the current active Salt Minion and update the Salt Minion there.
+
+If update is finished successfully, Salt Minion is restarted into a new environment, which is set as default. Previos environment is kept, and is purged over one update cycle (one after current update). This way administrator always has an ability to roll-back to the previous known working version.
+
+If update failed, Salt Minion keeps running without changes, report is created and new Virtual Environment is purged immeditately.
+
+As shown on the _Figure 2_, the difference between RPM-based updates is that the software components are overwritten, instead of installed nearby. Therefore once Salt Minion is broken on upgrade for whatever reasons, it might be no longer possible to revert this with the faulty Salt Minion itself.
+
+![Layout](images/00065-ve.png)
+
+_Figure 2_
+
+According to the _Figure 2_, if updated Salt Minion is successfully updated and is trusted enough, then the statuses of each virtual environment are shifted, and "Previous Salt Minion" is purged.
+
+Naturally, the same way it is possible to update the entire Python version on production and roll-back to the previous Python version, in case things go wrong.
+
+## Installation/Update Process
+
+This RFC is separating Python installation/update process from the Salt Minion installation/update processes, both on all architectures and OS versions. Consequently, mechanisms are different.
+
+### Step 1: Python Interpreter
+
+Python interpreter in "vanilla version" should be packaged as usual in RPM, per OS. However, compiled statically for that specific OS version and architecture. This alone allows update Python interpreter independently from the currently running Salt Minion. The diffrence is that the Python interpreter is installed in an optional namespace, e.g. `/opt/python<VERSION>/bin/python<VERSION>` path. The very RPM package, however, is served by Uyuni Server repository, instead of default channels.
+
+This RPM package remains being vendor supported.
+
+### Step 2: Virtual Environment
+
+Once Python interpreter is delivered and installed on a managed system, Virtual Environment is created. It should contain Package Installer for Python (PIP) and Salt Minion required packages list.
+
+Salt Minion is no longer installed from the RPM package, but bootstrapped by PIP environment. However, the distributed package version that is on the PyPi server remains vendor supported.
+
+> Idea for audit: checksums of all the files in the Virtual Environment can be calculated and stored in the database that might run on PyPi container. This way it is possible to detect if the environment was tampered.
```

Comment:
You need more than that: https://github.com/SUSE-Cloud/venvjail/blob/master/venvjail.py#L426

Reply:
Not for this paradigm.

Reply:
Ouch. So no L3 support.

Reply:
Not the way you imagine.

Reply:
That is fair. Maybe you can elaborate it in the doc?

Reply:
Yes, I will update the RFC, since it clearly missing many parts. Thanks for help!